### PR TITLE
treewide: Replace `EntityType` by `EntityState`

### PR DESCRIPTION
**Pull request by haslersn**
This is a rather big change and needs some review. The API changes however are minimal.

### CODE REVIEW

**Comment by st116290** [comment]
This would lead to a rather big time waste getting everything in our exercise sheets and Task/Solution classes up to date for minimal benefit. Is there a reason to merge this request before monday or can it wait until after the java introduction course is finished?

**Comment by haslersn** [comment]
The places where the tasks and solutions have to be changed are extremely easy to find via the compile errors. It took me 2 minutes to do all of the needed changes locally.

With this change, students can add Entity classes without modifying the ICGE.

**Comment by st116290** [comment]
That is a good point for Luigi. However the rendering and collisions still use hardcoded class checks. So entityState at least needs a method like isSolid() and one like getSpriteSet() or getSpriteId()

**Comment by haslersn** [comment]
Such stuff can be added in the future no matter whether we merge this PR or not.

Here is a patch for the current master of FIUS/JVK-2018@d014405d79f2ef1f7fd23829703eeeb2ad072404 :

[FIUS-ICGE-pull-37.txt](https://github.com/FIUS/ICGE/files/2453793/FIUS-ICGE-pull-37.txt)

**Comment by st116290** [comment]
But then we have exactly the same situation as with EntityState. We have to implement both the EntityState and the EntityClass for luigi for us to be able to show the correct sprites for luigi. We can do the same with the current EntityType and already have a finished pr for that.
As EntityType and EntityState rely on different concepts to work they look and feel quite different. The developers of the tasks would also need time to adapt to this change.

**Comment by haslersn** [comment]
In the tasks there was conceptually a single thing that changed: Instead of `EntityType.WALL`, use `WallState.class`. (Same for other `EntityType`s.)

The finished PR for Luigi just adds Luigi but the students cannot write Luigi themselves. If you'd remove the `Luigi` class from the PR, it wouldn't compile, since the `EntityType.LUIGI` in the `ICGE` uses `Luigi::new`. The only way to introduce new `Entity`s would be to prepare the corresponding `EntityType` in the `ICGE` **and also** create the `Entity`'s class with an empty constructor.

**Comment by st116290** [comment]
We could also change the EntityType enum to include a createEntity() method that does not work for luigi. As this method would only be called if luigi is set via a territory editor which we never do this would have nearly the same effects as this pr with minimal changes. Since the course starts on monday i would prefer minimal changes...

**Comment by haslersn** [comment]
Is there any problem with the PR besides that the tasks have to use `new WallState()` instead of `EntityType.WALL`?

Besides the possibility to add arbitrary new `Entity` classes, this PR also has the useful effect that entity state is never lost when using `setTerritory()`.

I updated the PR to use our newest master such that it merges fine.

Also, here's the patch for the current master of FIUS/JVK-2018@d014405d79f2ef1f7fd23829703eeeb2ad072404 which includes the necessary `pom.xml` change:

[FIUS-ICGE-pull-37.txt](https://github.com/FIUS/ICGE/files/2453820/FIUS-ICGE-pull-37.txt)

File: src/main/java/de/unistuttgart/informatik/fius/icge/territory/WorldObject.java [hunk 1]
```diff
@@ -13,39 +13,30 @@
...
+    public WorldObject(EntityState state, int column, int row, float z) {
+        this(state, column, row, z, Direction.EAST);
     }
 
-    public WorldObject(EntityType type, int column, int row) {
-        this(type, column, row, Direction.EAST);
-    }
-
-    @Override
-    public boolean equals(Object other) {
```
Comment (neumantm) [inline src/main/java/de/unistuttgart/informatik/fius/icge/territory/WorldObject.java]: Why remove the equals method?

Comment (haslersn) [reply src/main/java/de/unistuttgart/informatik/fius/icge/territory/WorldObject.java]: It didn't check for the equality of the state and I think the `equals` method should be deep if it exists.

**Pull request merged**